File: clients/android/app/src/main/java/com/syncbridge/android/localsend/LocalSendProtocol.kt
package com.syncbridge.android.localsend

import org.json.JSONArray
import org.json.JSONObject
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.UUID

object LocalSendProtocol {
    const val SERVICE_TYPE = "_syncbridge-localsend._tcp."
    const val CHUNK_SIZE = 4 * 1024 * 1024
    private val CHUNK_MAGIC = byteArrayOf(0x53, 0x42, 0x4C, 0x53) // SBLS

    data class FileEntry(
        val index: Int,
        val name: String,
        val relativePath: String,
        val size: Long,
    )

    data class Offer(
        val id: String,
        val sender: String,
        val files: List<FileEntry>,
    )

    data class Resume(
        val id: String,
        val offsets: Map<Int, Long>,
    )

    fun newTransferId(): String = UUID.randomUUID().toString()

    fun encodeOffer(offer: Offer): String {
        val files = JSONArray()
        offer.files.forEach { f ->
            files.put(
                JSONObject()
                    .put("index", f.index)
                    .put("name", f.name)
                    .put("relativePath", f.relativePath)
                    .put("size", f.size),
            )
        }
        return JSONObject()
            .put("op", "offer")
            .put("id", offer.id)
            .put("sender", offer.sender)
            .put("files", files)
            .toString() + "\n"
    }

    fun encodeAccept(id: String): String =
        JSONObject().put("op", "accept").put("id", id).toString() + "\n"

    fun encodeReject(id: String, reason: String): String =
        JSONObject().put("op", "reject").put("id", id).put("reason", reason).toString() + "\n"

    fun encodeFileBegin(id: String, index: Int, offset: Long): String =
        JSONObject()
            .put("op", "file_begin")
            .put("id", id)
            .put("index", index)
            .put("offset", offset)
            .toString() + "\n"

    fun encodeChunkAck(id: String, index: Int, offset: Long): String =
        JSONObject()
            .put("op", "chunk_ack")
            .put("id", id)
            .put("index", index)
            .put("offset", offset)
            .toString() + "\n"

    fun encodeFileEnd(id: String, index: Int): String =
        JSONObject().put("op", "file_end").put("id", id).put("index", index).toString() + "\n"

    fun encodeComplete(id: String): String =
        JSONObject().put("op", "complete").put("id", id).toString() + "\n"

    fun encodeResume(id: String, offsets: Map<Int, Long>): String {
        val arr = JSONArray()
        offsets.forEach { (idx, off) ->
            arr.put(JSONObject().put("index", idx).put("offset", off))
        }
        return JSONObject().put("op", "resume").put("id", id).put("files", arr).toString() + "\n"
    }

    fun parseMessage(line: String): JSONObject = JSONObject(line.trim())

    fun parseOffer(obj: JSONObject): Offer {
        val files = mutableListOf<FileEntry>()
        val arr = obj.getJSONArray("files")
        for (i in 0 until arr.length()) {
            val f = arr.getJSONObject(i)
            files.add(
                FileEntry(
                    index = f.getInt("index"),
                    name = f.getString("name"),
                    relativePath = f.getString("relativePath"),
                    size = f.getLong("size"),
                ),
            )
        }
        return Offer(
            id = obj.getString("id"),
            sender = obj.getString("sender"),
            files = files,
        )
    }

    fun parseResume(obj: JSONObject): Resume {
        val offsets = mutableMapOf<Int, Long>()
        val arr = obj.getJSONArray("files")
        for (i in 0 until arr.length()) {
            val f = arr.getJSONObject(i)
            offsets[f.getInt("index")] = f.getLong("offset")
        }
        return Resume(id = obj.getString("id"), offsets = offsets)
    }

    fun writeChunk(out: OutputStream, fileIndex: Int, offset: Long, data: ByteArray, length: Int) {
        val buf = ByteBuffer.allocate(18 + length).order(ByteOrder.BIG_ENDIAN)
        buf.put(CHUNK_MAGIC)
        buf.putShort(fileIndex.toShort())
        buf.putLong(offset)
        buf.putInt(length)
        buf.put(data, 0, length)
        out.write(buf.array())
        out.flush()
    }

    fun readChunk(`in`: InputStream): Chunk? {
        val header = ByteArray(18)
        var read = 0
        while (read < 18) {
            val n = `in`.read(header, read, 18 - read)
            if (n < 0) return null
            read += n
        }
        if (!header.copyOfRange(0, 4).contentEquals(CHUNK_MAGIC)) return null
        val buf = ByteBuffer.wrap(header, 4, 14).order(ByteOrder.BIG_ENDIAN)
        val fileIndex = buf.short.toInt() and 0xFFFF
        val offset = buf.long
        val length = buf.int
        val payload = ByteArray(length)
        DataInputStream(`in`).readFully(payload)
        return Chunk(fileIndex, offset, payload)
    }

    data class Chunk(val fileIndex: Int, val offset: Long, val data: ByteArray)
}

File: clients/ios/LocalSendProtocol.swift
// LocalSendProtocol.swift — P2P LAN transfer control + chunk framing (no cloud).

import Foundation

enum LocalSendProtocol {
    static let serviceType = "_syncbridge-localsend._tcp"
    static let chunkSize = 4 * 1024 * 1024
    private static let chunkMagic: [UInt8] = [0x53, 0x42, 0x4C, 0x53] // SBLS

    struct FileEntry: Codable {
        let index: Int
        let name: String
        let relativePath: String
        let size: Int64
    }

    struct Offer: Codable {
        let op: String
        let id: String
        let sender: String
        let files: [FileEntry]

        init(id: String, sender: String, files: [FileEntry]) {
            op = "offer"
            self.id = id
            self.sender = sender
            self.files = files
        }
    }

    struct Accept: Codable {
        let op: String
        let id: String
        init(id: String) { op = "accept"; self.id = id }
    }

    struct Reject: Codable {
        let op: String
        let id: String
        let reason: String
        init(id: String, reason: String) { op = "reject"; self.id = id; self.reason = reason }
    }

    struct FileBegin: Codable {
        let op: String
        let id: String
        let index: Int
        let offset: Int64
        init(id: String, index: Int, offset: Int64) {
            op = "file_begin"; self.id = id; self.index = index; self.offset = offset
        }
    }

    struct FileEnd: Codable {
        let op: String
        let id: String
        let index: Int
        init(id: String, index: Int) { op = "file_end"; self.id = id; self.index = index }
    }

    struct Complete: Codable {
        let op: String
        let id: String
        init(id: String) { op = "complete"; self.id = id }
    }

    struct Resume: Codable {
        struct Entry: Codable {
            let index: Int
            let offset: Int64
        }
        let op: String
        let id: String
        let files: [Entry]
        init(id: String, files: [Entry]) { op = "resume"; self.id = id; self.files = files }
    }

    static func encodeLine<T: Encodable>(_ value: T) throws -> Data {
        var data = try JSONEncoder().encode(value)
        data.append(0x0A)
        return data
    }

    static func parseOp(_ line: String) -> (op: String, json: [String: Any])? {
        guard let data = line.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let op = json["op"] as? String else { return nil }
        return (op, json)
    }

    static func resumeOffsets(from json: [String: Any]) -> [Int: Int64] {
        guard let files = json["files"] as? [[String: Any]] else { return [:] }
        var out: [Int: Int64] = [:]
        for f in files {
            if let idx = f["index"] as? Int, let off = f["offset"] as? Int64 {
                out[idx] = off
            }
        }
        return out
    }
}
